### DISCUSSION TITLE
Template as array key

### DISCUSSION THREAD

**Discussion by Baptouuuu** [vimeo/psalm#7901] [Q&A]
Is there a way to use a template to specify a key of an array like so : https://psalm.dev/r/7469115e9c ?

`toArray('i', 1)` => `array{i: int}`

**Comment by orklah** [comment]
You can have something similar like: https://psalm.dev/r/24318db73e but it seems Psalm choke on templates in offsets for a shape: https://psalm.dev/r/c1eafbc924. It considers `K` as an offset, which is weird

**Reply by AndrolGenhald** [reply]
Hadn't considered this use case before, but I've always done array-shapes as `array{somekey: int}` rather than `array{"somekey": int}`, probably because it's familiar from JavaScript and TypeScript. It would be nice to support templates here, but it would be a breaking change, and I'm not sure exactly how non-`literal-string` templates would work, I'd think maybe we'd want it to only consider templates that are `of literal-string`, since otherwise the array shape isn't possible anyway.

**Reply by Baptouuuu** [reply]
Indeed it should be `literal-string`

For a bit of context for why I want to do this: I'm trying to build objects to define an OpenAPI schema that can be used to validate a request and that returns an array shape of the payload (my goal is to define my types once instead of duplicating them between OpenAPI and manually specifying the array shape in my controllers). So I need the types to compose and using `array<K, V>` as suggested (even though it works for the simple case above) is not of much help.

Do you know if it would be complicated to implement this ?

**Reply by AndrolGenhald** [reply]
I opened #7906.

I'm not sure exactly how this would work for many keys, are you planning to do something like `$array += toArray($key, $value)` for each key? If you have a bunch of keys and values I think a simpler solution might be to improve `array_combine` to return array shapes when both arguments are array shapes (something I've wanted for some of my own code as well).

**Reply by Baptouuuu** [reply]
No I only need to define one key at a time and then use intersections. For example `array_merge(toArray('foo', 42), toArray('bar', '*'))`

**Reply by AndrolGenhald** [reply]
It looks like [both addition and `array_merge`](https://psalm.dev/r/bc51602c8a) work to combine array shapes, so all we need to do then is implement template keys.
